services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py

```diff
@@ -137,74 +135,51 @@ async def get_dynamic_sidecars_to_observe(
     ]
 
 
-async def _extract_task_data_from_service_for_state(
-    service_id: str,
-    dynamic_sidecar_settings: DynamicSidecarSettings,
-    target_statuses: set[str],
-) -> dict[str, Any]:
-    """Waits until the dynamic-sidecar task is in one of the target_statuses
-    and then returns the task"""
-
-    async def _sleep_or_error(started: float, task: dict):
-        await asyncio.sleep(1.0)
-        elapsed = time.time() - started
-        if (
-            elapsed
-            > dynamic_sidecar_settings.DYNAMIC_SIDECAR_TIMEOUT_FETCH_DYNAMIC_SIDECAR_NODE_ID
-        ):
-            raise DynamicSidecarError(
-                "Timed out while searching for an assigned NodeID for "
-                f"service_id={service_id}. Last task inspect result: {task}"
-            )
-
-    async with docker_client() as client:
-        service_state: Optional[str] = None
-        task: dict[str, Any] = {}
-
-        started = time.time()
+async def get_dynamic_sidecar_placement(
+    service_id: str, dynamic_sidecar_settings: DynamicSidecarSettings
+) -> str:
+    """
+    Waits until the service has a task in `running` state and
+    returns it's `docker_node_id`.
+    It is assumed that a `docker_node_id` exists if the service
+    is in `running` state.
+    """
 
-        while service_state not in target_statuses:
+    @retry(
+        wait=wait_random_exponential(multiplier=2, min=1, max=20),
+        stop=stop_after_delay(
+            dynamic_sidecar_settings.DYNAMIC_SIDECAR_TIMEOUT_FETCH_DYNAMIC_SIDECAR_NODE_ID
+        ),
+    )
+    async def _get_task_data_when_service_running(service_id: str) -> Mapping[str, Any]:
```

Comment:
You claim that

> Will reduce pressure on docker swarm when a lot of services are started together.

At first sight it looks to me that you are rewriting the same logic much more cleanly with tenacity. What is the key difference that I am missing and that reduces the "pressuse on the swarm"? 
thx

Reply:
You are correct, the code was rewritten with tenacity and results in being more clean.

The key here is `wait_random_exponential`. The part **random** is important. Please have look at [wait_random_exponential's docs](https://tenacity.readthedocs.io/en/latest/api.html#tenacity.wait.wait_random_exponential), they point out to [some an article from AWS](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) on the idea behind.

The idea behind it is to avoid having concurrent retrying calls at the same interval. If your system is failing and all your clients retry at the same time, it makes it difficult for the system to recover. Ideally you'd like to distribute the retries in time.

I will also add this as a comment, like you requested.